curp/src/client_new/tests.rs

```diff
@@ -0,0 +1,645 @@
+            .return_once(move |_req, _timeout| {
+                assert!(id == 0, "wait synced should send to leader");
+                assert!(
+                    flag_c.load(std::sync::atomic::Ordering::Relaxed),
+                    "fetch_leader should invoke first"
+                );
+                Ok(tonic::Response::new(WaitSyncedResponse::new_from_result::<
+                    TestCommand,
+                >(
+                    Ok(TestCommandResult::default()),
+                    Some(Ok(1.into())),
+                )))
+            });
+    });
+    let unary = Unary::<TestCommand>::new(connects, None, None);
+    let res = unary.slow_round(ProposeId(0, 0)).await.unwrap().unwrap();
+    assert_eq!(LogIndex::from(res.0), 1);
+    assert_eq!(res.1, TestCommandResult::default());
+}
+
+#[traced_test]
+#[tokio::test]
+async fn test_unary_propose_fast_path_works() {
+    let connects = init_mocked_connects(5, |id, conn| {
+        conn.expect_propose().return_once(move |_req, _timeout| {
+            let resp = match id {
+                0 => ProposeResponse::new_result::<TestCommand>(&Ok(TestCommandResult::default())),
+                1 | 2 | 3 => ProposeResponse::new_empty(),
+                4 => return Err(CurpError::key_conflict()),
+                _ => unreachable!("there are only 5 nodes"),
+            };
+            Ok(tonic::Response::new(resp))
+        });
+        conn.expect_wait_synced()
+            .return_once(move |_req, _timeout| {
+                assert!(id == 0, "wait synced should send to leader");
+                std::thread::sleep(Duration::from_millis(100));
+                Ok(tonic::Response::new(WaitSyncedResponse::new_from_result::<
+                    TestCommand,
+                >(
+                    Ok(TestCommandResult::default()),
+                    Some(Ok(1.into())),
+                )))
+            });
+    });
+    let unary = Unary::<TestCommand>::new(connects, None, Some((0, 1)));
+    let res = unary
+        .propose(&TestCommand::default(), true)
+        .await
+        .unwrap()
+        .unwrap();
+    assert_eq!(res, (TestCommandResult::default(), None));
+}
+
+#[traced_test]
+#[tokio::test]
+async fn test_unary_propose_slow_path_works() {
+    let connects = init_mocked_connects(5, |id, conn| {
+        conn.expect_propose().return_once(move |_req, _timeout| {
+            let resp = match id {
+                0 => ProposeResponse::new_result::<TestCommand>(&Ok(TestCommandResult::default())),
+                1 | 2 | 3 => ProposeResponse::new_empty(),
+                4 => return Err(CurpError::key_conflict()),
+                _ => unreachable!("there are only 5 nodes"),
+            };
+            Ok(tonic::Response::new(resp))
+        });
+        conn.expect_wait_synced()
+            .return_once(move |_req, _timeout| {
+                assert!(id == 0, "wait synced should send to leader");
+                std::thread::sleep(Duration::from_millis(100));
```

Comment:
It seems this will block the current thread and the assert below will always hold, maybe specify the runtime to multi threaded?

Reply:
No effect, as this is not a block that can be suspended(i.e. no `.await`). There is always only one task running in the entrie test case. 

It seems that we cannot pass an async block into `return_once`

The sleep here implies that we did indeed execute wait sync.